### ISSUE TITLE
Pulumi 3.25.0 breaks pulumi_aws.iam.get_policy_document with explicit AWS provider

### ISSUE THREAD

**Issue by piotr-bzdyl-vertexinc** [pulumi/pulumi#9060]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

Pulumi 3.25.0
Runtime: python
Pulumi AWS: pulumi_aws-4.37.5

```
$ cat Pulumi.prod.yaml
config:
  pulumi:disable-default-providers:
    - aws

```

And I have the code like the following (simplified):

```
aws_provider = ...
print(aws_provider)
policy = pulumi_aws.iam.get_policy_document(
    statements=[
        pulumi_aws.iam.GetPolicyDocumentStatementArgs(
            actions=['s3:*'],
        ),
    ],
    opts=pulumi.InvokeOptions(provider=aws_provider),
)

```

`print(aws_provider)` result:
```
<pulumi_aws.provider.Provider object at 0x7f559c72d940>
```

and an error is raised:
```
Exception: invoke of aws:iam/getPolicyDocument:getPolicyDocument failed: Default provider for 'aws' disabled. 'aws:iam/getPolicyDocument:getPolicyDocument' must use an explicit provider.
```

When I just change the Pulumi version to 3.24.1 (with no other changes in the code), the issue disappears.

**Comment by stack72** [comment]
Hey @AaronFriel this feels related to something you have been working on recently?

**Comment by AaronFriel** [comment]
Bisected this to this commit:

93f526d4afa158aa85b6dfdeea5017724cbd1d40

**Comment by AaronFriel** [comment]
@piotr-bzdyl-vertexinc we've got a fix under way in #9067, we're just adding tests before we merge.

In the meantime the workaround is to not use "disable-default-providers", though if that's not an option, we expect to have a point release out today addressing the root cause.

**Comment by AaronFriel** [comment]
See:
- #9055
